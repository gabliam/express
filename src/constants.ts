/** Type in registry */
export const TYPE = {
  Controller: 'ControllerType'
};

export const METADATA_KEY = {
  controller: '_controller',
  controllerMethod: '_controller-method',
  controllerParameter: '_controller-parameter',
  middleware: '_middleware',
  MiddlewareConfig: '_middlewareConfig',
  MiddlewareErrorConfig: '_middlewareErrorConfig'
};

export enum PARAMETER_TYPE {
  REQUEST,
  RESPONSE,
  PARAMS,
  QUERY,
  BODY,
  HEADERS,
  COOKIES,
  NEXT
}

export const DEFAULT_ROUTING_ROOT_PATH = '/';

export const EXPRESS_PLUGIN_CONFIG = Symbol('GABLIAM/EXPRESS_PLUGIN_CONFIG');

export const APP = Symbol('GABLIAM/EXPRESS_APP');
export const SERVER = Symbol('GABLIAM/EXPRESS_SERVER');
