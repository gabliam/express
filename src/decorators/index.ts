export * from './controller';
export * from './middleware';
export * from './config';
export * from './method';
export * from './params';
