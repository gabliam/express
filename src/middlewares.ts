import { ExpressConfig } from './interfaces';

export class MiddlewareConfig {
  private _middlewares: ExpressConfig[] = [];

  private _errorMiddlewares: ExpressConfig[] = [];


  addMiddleware(middleware: ExpressConfig) {
    this._middlewares.push(middleware);
  }

  addErrorMiddleware(middleware: ExpressConfig) {
    this._errorMiddlewares.push(middleware);
  }

  get middlewares() {
    return this._middlewares.slice();
  }

  get errorMiddlewares() {
    return this._errorMiddlewares.slice();
  }
}
