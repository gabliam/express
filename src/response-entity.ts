import { OutgoingHttpHeaders } from 'http';
import * as HttpStatus from 'http-status-codes';

export class ResponseEntity<T = any> {
  private _status = HttpStatus.OK;

  private _headers: OutgoingHttpHeaders = {};

  private _body: T | null = null;

  constructor(body: T, status: number = 200) {
    this._body = body;
    this._status = status;
  }

  get status() {
    return this._status;
  }

  get headers() {
    return this._headers;
  }

  get body() {
    return this._body;
  }

  hasHeader() {
    return Object.keys(this._headers).length !== 0;
  }

  addHeader(headerName: string, value: any) {
    this._headers[headerName] = value;
  }
}

/**
 * A shortcut for creating a ResponseEntity with the given body and the status set to OK.
 * @param body
 */
export function ok<T = any>(body?: T) {
  return new ResponseEntity(body, HttpStatus.OK);
}

/**
 * Create a ResponseEntity with an ACCEPTED status.
 */
export function accepted() {
  return new ResponseEntity(null, HttpStatus.ACCEPTED);
}

/**
 * Create a ResponseEntity with a BAD_REQUEST status.
 */
export function badRequest() {
  return new ResponseEntity(null, HttpStatus.BAD_REQUEST);
}

/**
 * Create a ResponseEntity with a NO_CONTENT status.
 */
export function noContent() {
  return new ResponseEntity(null, HttpStatus.NO_CONTENT);
}

/**
 * Create a ResponseEntity with a NOT_FOUND status.
 */
export function notFound() {
  return new ResponseEntity(null, HttpStatus.NOT_FOUND);
}
