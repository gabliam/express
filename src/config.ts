import { PluginConfig, Bean, Value } from '@gabliam/core';
import { DEFAULT_ROUTING_ROOT_PATH, EXPRESS_PLUGIN_CONFIG } from './constants';
import { ExpressPluginConfig } from './interfaces';
import * as Joi from 'joi';

@PluginConfig()
export class RestPluginConfig {
    @Value('application.express.rootPath', Joi.string())
    rootPath = DEFAULT_ROUTING_ROOT_PATH;

    @Value('application.express.port', Joi.number().positive())
    port: number = process.env.PORT || 3000;

    @Value('application.express.hostname', Joi.string())
    hostname: string;


    @Bean(EXPRESS_PLUGIN_CONFIG)
    restConfig(): ExpressPluginConfig {
        return {
            rootPath: this.rootPath,
            port: this.port,
            hostname: this.hostname
        };
    }
}
