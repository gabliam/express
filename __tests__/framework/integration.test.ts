import {
  Controller,
  Get,
  Post,
  Put,
  Patch,
  Head,
  Delete,
  Method
} from '../../src/index';
import * as e from 'express';
import { ExpressPluginTest } from '../express-plugin-test';
import * as supertest from 'supertest';
import { Config, Bean } from '@gabliam/core';
import { CUSTOM_ROUTER_CREATOR } from '../../src/constants';

let appTest: ExpressPluginTest;

beforeEach(async () => {
  appTest = new ExpressPluginTest();
});

afterEach(async () => {
  await appTest.destroy();
});

describe('Integration Tests:', () => {
  test('should work for async controller methods', async () => {
    @Controller('/')
    class TestController {
      @Get('/')
      async getTest(req: e.Request, res: e.Response) {
        return new Promise(resolve => {
          setTimeout(resolve, 100, 'GET');
        });
      }
    }
    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app).get('/').expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should work for async controller methods that fails', async () => {
    @Controller('/')
    class TestController {
      @Get('/')
      public getTest(req: e.Request, res: e.Response) {
        return new Promise((resolve, reject) => {
          setTimeout(reject, 100, 'GET');
        });
      }
    }
    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app).get('/').expect(500);
    expect(response).toMatchSnapshot();
  });

  test('should work for methods which call nextFunc()', async () => {
    @Controller('/')
    class TestController {
      @Get('/')
      public getTest(
        req: e.Request,
        res: e.Response,
        nextFunc: e.NextFunction
      ) {
        nextFunc();
      }

      @Get('/')
      public getTest2(req: e.Request, res: e.Response) {
        return 'GET';
      }
    }

    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app).get('/').expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should work for async methods which call nextFunc()', async () => {
    @Controller('/')
    class TestController {
      @Get('/')
      public getTest(
        req: e.Request,
        res: e.Response,
        nextFunc: e.NextFunction
      ) {
        return new Promise(resolve => {
          setTimeout(
            () => {
              nextFunc();
              resolve();
            },
            100,
            'GET'
          );
        });
      }

      @Get('/')
      public getTest2(req: e.Request, res: e.Response) {
        return 'GET';
      }
    }
    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app).get('/').expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should work for async methods called by nextFunc()', async () => {
    @Controller('/')
    class TestController {
      @Get('/')
      public getTest(
        req: e.Request,
        res: e.Response,
        nextFunc: e.NextFunction
      ) {
        nextFunc();
      }

      @Get('/')
      public getTest2(req: e.Request, res: e.Response) {
        return new Promise(resolve => {
          setTimeout(resolve, 100, 'GET');
        });
      }
    }
    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app).get('/').expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should work for each shortcut decorator', async () => {
    @Controller('/')
    class TestController {
      @Get('/')
      public getTest(req: e.Request, res: e.Response) {
        res.send('GET');
      }
      @Post('/')
      public postTest(req: e.Request, res: e.Response) {
        res.send('POST');
      }
      @Put('/')
      public putTest(req: e.Request, res: e.Response) {
        res.send('PUT');
      }
      @Patch('/')
      public patchTest(req: e.Request, res: e.Response) {
        res.send('PATCH');
      }
      @Head('/')
      public headTest(req: e.Request, res: e.Response) {
        res.send('HEAD');
      }
      @Delete('/')
      public deleteTest(req: e.Request, res: e.Response) {
        res.send('DELETE');
      }
    }
    appTest.addClass(TestController);
    await appTest.build();
    const agent = supertest(appTest.app);

    const rd = await agent.delete('/').expect(200);
    expect(rd).toMatchSnapshot();

    const rh = await agent.head('/').expect(200);
    expect(rh).toMatchSnapshot();

    const rpatch = await agent.patch('/').expect(200);
    expect(rpatch).toMatchSnapshot();

    const rput = await agent.put('/').expect(200);
    expect(rput).toMatchSnapshot();

    const rpost = await agent.post('/').expect(200);
    expect(rpost).toMatchSnapshot();

    const rget = await agent.get('/').expect(200);
    expect(rget).toMatchSnapshot();
  });

  test('should work for more obscure HTTP methods using the httpMethod decorator', async () => {
    @Controller('/')
    class TestController {
      @Method('propfind', '/')
      public getTest(req: e.Request, res: e.Response) {
        res.send('PROPFIND');
      }
    }

    appTest.addClass(TestController);
    await appTest.build();

    const response = await supertest(appTest.app).propfind('/').expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should use returned values as response', async () => {
    @Controller('/')
    class TestController {
      @Get('/')
      public getTest(req: e.Request, res: e.Response) {
        return { hello: 'world' };
      }
    }

    appTest.addClass(TestController);
    await appTest.build();

    const response = await supertest(appTest.app).get('/').expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should use custom router passed from configuration', async () => {
    @Controller('/CaseSensitive')
    class TestController {
      @Get('/Endpoint')
      public get() {
        return 'Such Text';
      }
    }

    @Config()
    class Conf {
      @Bean(CUSTOM_ROUTER_CREATOR)
      custom() {
        return () =>
          e.Router({
            caseSensitive: true
          });
      }
    }
    appTest.addClass(Conf);
    appTest.addClass(TestController);
    await appTest.build();

    const agent = supertest(appTest.app);

    const expectedSuccess = await agent
      .get('/CaseSensitive/Endpoint')
      .expect(200, 'Such Text');

    expect(expectedSuccess).toMatchSnapshot();

    const expectedNotFound1 = await agent
      .get('/casesensitive/endpoint')
      .expect(404);

    expect(expectedNotFound1).toMatchSnapshot();

    const expectedNotFound2 = await agent
      .get('/CaseSensitive/endpoint')
      .expect(404);

    expect(expectedNotFound2).toMatchSnapshot();
  });

  test('should use custom routing configuration', async () => {
    @Controller('/ping')
    class TestController {
      @Get('/endpoint')
      public get() {
        return 'pong';
      }
    }

    appTest.addClass(TestController);
    appTest.addConf('application.express.rootPath', '/api/v1');
    await appTest.build();

    const response = await supertest(appTest.app)
      .get('/api/v1/ping/endpoint')
      .expect(200);

    expect(response).toMatchSnapshot();
  });
});
