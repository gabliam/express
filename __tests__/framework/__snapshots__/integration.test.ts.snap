// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration Tests: decorator Controller should use custom router passed from configuration 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "9",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"9-pQ/Pu6+qfOrXcJtZvB+frrlxk/c\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "Such Text",
}
`;

exports[`Integration Tests: decorator Controller should use custom router passed from configuration 2`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "161",
    "content-security-policy": "default-src 'self'",
    "content-type": "text/html; charset=utf-8",
    "x-content-type-options": "nosniff",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 404,
  "text": "<!DOCTYPE html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<title>Error</title>
</head>
<body>
<pre>Cannot GET /casesensitive/endpoint</pre>
</body>
</html>
",
}
`;

exports[`Integration Tests: decorator Controller should use custom router passed from configuration 3`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "161",
    "content-security-policy": "default-src 'self'",
    "content-type": "text/html; charset=utf-8",
    "x-content-type-options": "nosniff",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 404,
  "text": "<!DOCTYPE html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<title>Error</title>
</head>
<body>
<pre>Cannot GET /CaseSensitive/endpoint</pre>
</body>
</html>
",
}
`;

exports[`Integration Tests: decorator Controller should use custom routing configuration 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"4-DlFKBmK8tp3IY5U9HOJuPUDoGoc\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "pong",
}
`;

exports[`Integration Tests: decorator Controller should use returned values as response 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "17",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"11-IkjuL6CqqtmReFMfkkvwC0sKj04\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"hello\\":\\"world\\"}",
}
`;

exports[`Integration Tests: decorator Controller should work for async controller methods 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Integration Tests: decorator Controller should work for async controller methods that fails 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "130",
    "content-security-policy": "default-src 'self'",
    "content-type": "text/html; charset=utf-8",
    "x-content-type-options": "nosniff",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 500,
  "text": "<!DOCTYPE html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<title>Error</title>
</head>
<body>
<pre>GET</pre>
</body>
</html>
",
}
`;

exports[`Integration Tests: decorator Controller should work for async methods called by nextFunc() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Integration Tests: decorator Controller should work for async methods which call nextFunc() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Integration Tests: decorator Controller should work for each shortcut decorator 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "6",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"6-1vVjYJjNRYzp0ik5+OPo3qsOm9A\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "DELETE",
  },
  "status": 200,
  "text": "DELETE",
}
`;

exports[`Integration Tests: decorator Controller should work for each shortcut decorator 2`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "HEAD",
  },
  "status": 200,
  "text": undefined,
}
`;

exports[`Integration Tests: decorator Controller should work for each shortcut decorator 3`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"5-n74p1Ndux1ZCGu4G4PNPF/pIj/M\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "PATCH",
  },
  "status": 200,
  "text": "PATCH",
}
`;

exports[`Integration Tests: decorator Controller should work for each shortcut decorator 4`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-CRsM5C6wvZYWnqALFt2Tj21jrJU\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "PUT",
  },
  "status": 200,
  "text": "PUT",
}
`;

exports[`Integration Tests: decorator Controller should work for each shortcut decorator 5`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"4-Yf+Bwwqjx254r+pisuO9HfpJ6FQ\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "POST",
  },
  "status": 200,
  "text": "POST",
}
`;

exports[`Integration Tests: decorator Controller should work for each shortcut decorator 6`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Integration Tests: decorator Controller should work for methods which call nextFunc() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Integration Tests: decorator Controller should work for more obscure HTTP methods using the httpMethod decorator 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "8",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"8-q7JH/ElbTfQ5LDMKp+kb5Bpx9So\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "PROPFIND",
  },
  "status": 200,
  "text": "PROPFIND",
}
`;

exports[`Integration Tests: decorator RestController should use custom router passed from configuration 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "11",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"b-vqDL+6YwrE3yJSTRBCWcavKRhdk\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "\\"Such Text\\"",
}
`;

exports[`Integration Tests: decorator RestController should use custom router passed from configuration 2`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "161",
    "content-security-policy": "default-src 'self'",
    "content-type": "text/html; charset=utf-8",
    "x-content-type-options": "nosniff",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 404,
  "text": "<!DOCTYPE html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<title>Error</title>
</head>
<body>
<pre>Cannot GET /casesensitive/endpoint</pre>
</body>
</html>
",
}
`;

exports[`Integration Tests: decorator RestController should use custom router passed from configuration 3`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "161",
    "content-security-policy": "default-src 'self'",
    "content-type": "text/html; charset=utf-8",
    "x-content-type-options": "nosniff",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 404,
  "text": "<!DOCTYPE html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<title>Error</title>
</head>
<body>
<pre>Cannot GET /CaseSensitive/endpoint</pre>
</body>
</html>
",
}
`;

exports[`Integration Tests: decorator RestController should use custom routing configuration 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "6",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"6-uBnwlsJiQ3kuZAgKKWB4aV5ugdE\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "\\"pong\\"",
}
`;

exports[`Integration Tests: decorator RestController should use returned values as response 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "17",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"11-IkjuL6CqqtmReFMfkkvwC0sKj04\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "{\\"hello\\":\\"world\\"}",
}
`;

exports[`Integration Tests: decorator RestController should work for async controller methods 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"5-F+E/6Ma6IsYxNfQFCboz97TKFWQ\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "\\"GET\\"",
}
`;

exports[`Integration Tests: decorator RestController should work for async controller methods that fails 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "130",
    "content-security-policy": "default-src 'self'",
    "content-type": "text/html; charset=utf-8",
    "x-content-type-options": "nosniff",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 500,
  "text": "<!DOCTYPE html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<title>Error</title>
</head>
<body>
<pre>GET</pre>
</body>
</html>
",
}
`;

exports[`Integration Tests: decorator RestController should work for async methods called by nextFunc() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"5-F+E/6Ma6IsYxNfQFCboz97TKFWQ\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "\\"GET\\"",
}
`;

exports[`Integration Tests: decorator RestController should work for async methods which call nextFunc() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"5-F+E/6Ma6IsYxNfQFCboz97TKFWQ\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "\\"GET\\"",
}
`;

exports[`Integration Tests: decorator RestController should work for each shortcut decorator 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "6",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"6-1vVjYJjNRYzp0ik5+OPo3qsOm9A\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "DELETE",
  },
  "status": 200,
  "text": "DELETE",
}
`;

exports[`Integration Tests: decorator RestController should work for each shortcut decorator 2`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "HEAD",
  },
  "status": 200,
  "text": undefined,
}
`;

exports[`Integration Tests: decorator RestController should work for each shortcut decorator 3`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"5-n74p1Ndux1ZCGu4G4PNPF/pIj/M\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "PATCH",
  },
  "status": 200,
  "text": "PATCH",
}
`;

exports[`Integration Tests: decorator RestController should work for each shortcut decorator 4`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-CRsM5C6wvZYWnqALFt2Tj21jrJU\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "PUT",
  },
  "status": 200,
  "text": "PUT",
}
`;

exports[`Integration Tests: decorator RestController should work for each shortcut decorator 5`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "4",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"4-Yf+Bwwqjx254r+pisuO9HfpJ6FQ\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "POST",
  },
  "status": 200,
  "text": "POST",
}
`;

exports[`Integration Tests: decorator RestController should work for each shortcut decorator 6`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Integration Tests: decorator RestController should work for methods which call nextFunc() 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"5-F+E/6Ma6IsYxNfQFCboz97TKFWQ\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "\\"GET\\"",
}
`;

exports[`Integration Tests: decorator RestController should work for more obscure HTTP methods using the httpMethod decorator 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "8",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"8-q7JH/ElbTfQ5LDMKp+kb5Bpx9So\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "PROPFIND",
  },
  "status": 200,
  "text": "PROPFIND",
}
`;
